keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1416 @@
+not need to figure out when to remove `CSIStorageCapacity` objects
+created on another node. It is also better than the node as owner
+because a driver might no longer be needed on a node although the node
+continues to exist (for example, when deploying a driver to nodes is
+controlled via node labels).
+
+While external-provisioner runs, it needs to update and potentially
+delete `CSIStorageCapacity` objects:
+- when nodes change (for central provisioning)
+- when storage classes change (for persistent volumes)
+- when volumes were created or deleted (for central provisioning)
+- when volumes are resized or snapshots are created or deleted (for persistent volumes)
+- periodically, to detect changes in the underlying backing store (all cases)
+
+Because sidecars are currently separated, external-provisioner is
+unaware of resizing and snapshotting. The periodic polling will catch up
+with changes caused by those operations.
+
+### Using capacity information
+
+The Kubernetes scheduler already has a component, the [volume
+scheduling
+library](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/scheduling),
+which implements [topology-aware
+scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md).
+
+The
+[CheckVolumeBinding](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md#integrating-volume-binding-with-pod-scheduling)
+function gets extended to not only check for a compatible topology of
+a node (as it does now), but also to verify whether the node falls
+into a topology segment that has enough capacity left. This check is
+only necessary for PVCs that have not been bound yet.
+
+The lookup sequence will be:
+- find the `CSIDriver` object for the driver
+- check whether it has `CSIDriver.spec.storageCapacity` enabled
+- find all `CSIStorageCapacity` objects that have the right spec
+  (driver, accessible by node, storage class) and sufficient capacity.
+
+The specified volume size is compared against `Capacity` if
+available. A topology segment which has no reported capacity or a
+capacity that is too small is considered unusable at the moment and
+ignored.
+
+Each volume gets checked separately, independently of other volumes
+that are needed by the current pod or by volumes that are about to be
+created for other pods. Those scenarios remain problematic.
+
+Trying to model how different volumes affect capacity would be
+difficult. If the capacity represents "maximum volume size 10GiB", it may be possible
+to create exactly one such volume or several, so rejecting the
+pool after one volume could be a false negative. With "available
+capacity 10GiB" it may or may not be possible to create two volumes of
+5GiB each, so accepting the node for two such volumes could be a false
+positive.
+
+More promising might be to add prioritization of nodes based on how
```

Comment:
By saying "prioritization of nodes", it seems that you are talking about local storage here?

Reply:
Networked storage might also be available only on a subset of the cluster. So prioritization may also be applicable there.